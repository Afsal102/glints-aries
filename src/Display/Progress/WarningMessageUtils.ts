import { capitalize } from 'lodash';

const generateInvalidPropValueMessage = (component: string) =>
  `Invalid prop value on ${capitalize(component)} component: `;

const generateExpectedAndActualPropMessage = (
  propName: string,
  expectedPropTypeAndValue: string,
  actualProp: string
) =>
  `${propName} prop expected a ${expectedPropTypeAndValue} but\ received ${actualProp} instead. `;

const generateResolvedPropMessage = (resolvedPropValue: string) =>
  `We resolved the value to ${resolvedPropValue} to ensure it continues working as expected.\ `;

const generateComponentDocumentationMessage = (documentationURL: string) =>
  `Visit ${documentationURL} for more information on the components' behaviour.`;

const generateWarningMessage = (messageProps: IMessageProps) => {
  return (
    generateInvalidPropValueMessage(messageProps.component) +
    generateExpectedAndActualPropMessage(
      messageProps.propName,
      messageProps.expectedPropTypeAndValue,
      messageProps.actualProp
    ) +
    generateResolvedPropMessage(messageProps.resolvedPropValue) +
    generateComponentDocumentationMessage(messageProps.documentationURL)
  );
};

const makeShowWarningMessage = (defaultMessageProps: IDefaultMessageProps) => (
  messageProps: IMessageProps
) =>
  console.warn(
    generateWarningMessage({ ...messageProps, ...defaultMessageProps })
  );

const makeReducerToPairWarningMessagesWithShowMessageFn = (
  defaultMessageProps: IDefaultMessageProps
) => {
  const showWarningMessageWithDefaultProps = makeShowWarningMessage(
    defaultMessageProps
  );
  const reducer = (accumulator: object, currentWarningMessage: string) => ({
    ...accumulator,
    [currentWarningMessage]: showWarningMessageWithDefaultProps,
  });
  return reducer;
};

export const generateWarningMessages = (
  warningMessageTypes: string[],
  defaultWarningMessageProps: IDefaultMessageProps
) => {
  const pairWarningMessagesWithShowMessageFn = makeReducerToPairWarningMessagesWithShowMessageFn(
    defaultWarningMessageProps
  );
  return warningMessageTypes.reduce(pairWarningMessagesWithShowMessageFn, {});
};

interface IMessageProps {
  component: string;
  documentationURL: string;
  propName: string;
  expectedPropTypeAndValue: string;
  actualProp: any;
  resolvedPropValue: any;
}

interface IDefaultMessageProps {
  component: string;
  documentationURL: string;
  propName?: string;
  expectedPropTypeAndValue?: string;
  actualProp?: any;
  resolvedPropValue?: any;
}
