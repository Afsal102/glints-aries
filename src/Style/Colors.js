export const PrimaryColor = {
  glintsred: '#EC272B',
  glintsblue: '#0BAEEC',
  glintsyellow: '#FFF240',
};

export const SecondaryColor = {
  white: '#FFFFFF',
  whitesmoke: '#F5F5F5',
  black: '#000000',
  lightblack: '#777777',
  grey: '#646464',
  lightgrey: '#C6C6C6',
  buttercup: '#F5A623',
  purplepizza: '#FF00E5',
  blue: '#0869C2',
  bluegem: '#9013FE',
};

