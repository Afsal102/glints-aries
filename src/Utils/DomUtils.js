// DOM utilities

/**
 * Returns a value when ESC button is triggered.
 *
 * @param  {Boolean} value
 * @return {Boolean}
 */
export function escEvent(value) {
  const listener = (e) => {
    if (e.keyCode === 27) {
      return value();
    }
    return 0;
  };

  return listener;
}
