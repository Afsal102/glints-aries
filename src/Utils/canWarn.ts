export const canWarn =
  typeof console !== 'undefined' && process.env.NODE_ENV === 'development';
