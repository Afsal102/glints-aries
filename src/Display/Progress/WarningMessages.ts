import { generateWarningMessages } from './WarningMessageUtils';

const defaultWarningMessageProps = {
  component: 'progress',
  documentationURL:
    'http://glints-aries.surge.sh/?selectedKind=Display&selectedStory=Progress&full=0',
};

const warningMessageTypes = [
  'percentageValueOutsideRange',
  'percentageTypeInvalid',
  'percentageRangeExceedsValidLength',
  'sizeValueInvalid',
];

export const warningMessages = generateWarningMessages(
  warningMessageTypes,
  defaultWarningMessageProps
) as ProgressWarningMessageTypes;

interface ProgressWarningMessageTypes {
  percentageValueOutsideRange: Function;
  percentageTypeInvalid: Function;
  percentageRangeExceedsValidLength: Function;
  sizeValueInvalid: Function;
}
