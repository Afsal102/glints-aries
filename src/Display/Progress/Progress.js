/* @flow */

import React, { Fragment } from 'react';

import { ProgressContainer, ProgressLabelWrapper } from '../../Style/Display/ProgressStyle';
import { SecondaryColor } from '../../Style/Colors';

const Progress = (props: Props) => {
  const {
    className,
    percentage,
    ...defaultProps
  } = props;

  const progressValue = percentage > 100 ? 282.6 * (1 - (100 / 100)) : 282.6 * (1 - (percentage / 100));
  const color = percentage > 50 ? SecondaryColor.green : SecondaryColor.orange;

  return (
    <Fragment>
      {percentage >= 0 && (
        <ProgressContainer className={className} progress={progressValue} {...defaultProps}>
          <svg width="8em" height="8em" viewBox="0 0 100 100">
            <circle cx="50" cy="50" r="45" fill="none" stroke={SecondaryColor.lighterblack} strokeWidth="8" />
            <circle className="progress-circle__value" cx="50" cy="50" r="45" fill="none" stroke={color} strokeWidth="8" />
          </svg>
          <ProgressLabelWrapper>
            {`${percentage > 100 ? 100 : percentage}%`}
            <p>COMPLETE</p>
          </ProgressLabelWrapper>
        </ProgressContainer>
      )}
    </Fragment>
  );
};

type Props = {
  className: string,
  percentage: number,
}

export default Progress;
