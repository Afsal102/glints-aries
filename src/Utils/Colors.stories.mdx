import { ColorItem, ColorPalette, Meta } from '@storybook/addon-docs';
import { PrimaryColor, Greyscale, SecondaryColor } from './Colors';
import { getReadableColor } from './getReadableColor';
import { ColorSampler, ColorSample } from './ColorsStoriesHelpers';
import blue from './@next/colors/blue';
import green from './@next/colors/green';
import neutral from './@next/colors/neutral';
import orange from './@next/colors/orange';
import red from './@next/colors/red';
import yellow from './@next/colors/yellow';

<Meta title="Utilities/Colors" />

# Color

```
import { PrimaryColor, Greyscale } from 'glints-aries';

const MyComponent = styled.div`
  background-color: ${PrimaryColor.glintsyellow};
  color: ${PrimaryColor.grey};
`;
```

## PrimaryColor

<ColorSampler>
  {Object.keys(PrimaryColor).map(colorName => (
    <ColorSample
      key={colorName}
      bgColor={PrimaryColor[colorName]}
      fontColor={getReadableColor(PrimaryColor[colorName])}
      name={colorName}
    />
  ))}
</ColorSampler>

## Greyscale

<ColorSampler>
  {Object.keys(Greyscale).map(colorName => (
    <ColorSample
      key={colorName}
      bgColor={Greyscale[colorName]}
      fontColor={getReadableColor(Greyscale[colorName])}
      name={colorName}
    />
  ))}
</ColorSampler>

## SecondaryColor

<ColorSampler>
  {Object.keys(SecondaryColor).map(colorName => (
    <ColorSample
      key={colorName}
      bgColor={SecondaryColor[colorName]}
      fontColor={getReadableColor(SecondaryColor[colorName])}
      name={colorName}
    />
  ))}
</ColorSampler>

# @next

## Brand Colors

<ColorSampler>
  <ColorSample key={yellow.Brand} bgColor={yellow.Brand} name="yellow.Brand" />
  <ColorSample key={red.Brand} bgColor={red.Brand} name="red.Brand" />
  <ColorSample key={blue.Brand} bgColor={blue.Brand} name="blue.Brand" />
  <ColorSample key={orange.Brand} bgColor={orange.Brand} name="orange.Brand" />
</ColorSampler>

## Product Colors

<ColorPalette>
  <ColorItem
    title="blue"
    subtitle="blue.codeName"
    colors={Object.fromEntries(Object.entries(blue).slice(1, 7))}
  />
  <ColorItem title="green" subtitle="green.codeName" colors={green} />
  <ColorItem
    title="neutral whites"
    subtitle="neutral.codeName"
    colors={Object.fromEntries(Object.entries(neutral).slice(0, 5))}
  />
  <ColorItem
    title="neutral blacks"
    subtitle="neutral.codeName"
    colors={Object.fromEntries(Object.entries(neutral).slice(5, 10))}
  />
  <ColorItem
    title="orange"
    subtitle="orange.codeName"
    colors={Object.fromEntries(Object.entries(orange).slice(1, 6))}
  />
  <ColorItem
    title="red"
    subtitle="red.codeName"
    colors={Object.fromEntries(Object.entries(red).slice(1, 6))}
  />
  <ColorItem
    title="yellow"
    subtitle="yellow.codeName"
    colors={Object.fromEntries(Object.entries(yellow).slice(1, 6))}
  />
</ColorPalette>
