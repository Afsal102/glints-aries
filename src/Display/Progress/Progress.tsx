import * as React from 'react';
import { isNumber } from 'lodash';
import classNames from 'classnames';

import {
  ProgressContainer,
  ProgressContent,
  ProgressLabelWrapper,
  PercentageCompletion,
  LabelText,
} from '../../Style/Display/ProgressStyle';
import { PrimaryColor, SecondaryColor } from '../../Style/Colors';
import { warningMessages } from './WarningMessages';

const Progress: React.FunctionComponent<Props> = props => {
  const {
    className,
    percentage,
    percentageRange = [50],
    size = 8,
    content,
    ...defaultProps
  } = props;

  let normalizedPercentage;
  let normalizedSize;
  const isPercentageWithinCorrectRange =
    isNumber(percentage) && (percentage < 0 || percentage > 100);

  if (isPercentageWithinCorrectRange) {
    if (percentage < 0) {
      normalizedPercentage = 0;
    } else {
      normalizedPercentage = 100;
    }
    warningMessages.percentageValueOutsideRange({
      propName: 'percentage',
      expectedPropTypeAndValue: 'number between 0-100',
      actualProp: percentage,
      resolvedPropValue: normalizedPercentage,
    });
  } else if (!isNumber(percentage)) {
    normalizedPercentage = 0;
    warningMessages.percentageTypeInvalid({
      propName: 'percentage',
      expectedPropTypeAndValue: 'number between 0-100',
      actualProp: percentage,
      resolvedPropValue: normalizedPercentage,
    });
  } else {
    normalizedPercentage = percentage;
  }
  if (percentageRange.length > 2) {
    warningMessages.percentageRangeExceedsValidLength({
      propName: 'percentageRange',
      expectedPropTypeAndValue: 'number between 0-100',
      actualProp: percentageRange,
      resolvedPropValue: [percentageRange[0], percentageRange[1]],
    });
  }
  if (size < 1 || size > 10) {
    normalizedSize = size < 1 ? 1 : size > 10 ? 10 : size;
    warningMessages.percentageRangeExceedsValidLength({
      propName: 'size',
      expectedPropTypeAndValue: 'number between 1-10',
      actualProp: size,
      resolvedPropValue: normalizedSize,
    });
  } else {
    normalizedSize = size;
  }

  const progressValue =
    normalizedPercentage > 100
      ? 282.6 * (1 - 100 / 100)
      : 282.6 * (1 - normalizedPercentage / 100);
  const sizeInEm = (normalizedSize + 2) / 10;
  let color;

  if (percentageRange.length >= 2) {
    const [firstRange, secondRange] = percentageRange;
    color =
      normalizedPercentage > secondRange
        ? SecondaryColor.green
        : normalizedPercentage > firstRange
        ? SecondaryColor.orange
        : PrimaryColor.glintsred;
  } else if (percentageRange.length === 1) {
    const [firstRange] = percentageRange;
    color =
      normalizedPercentage > firstRange
        ? SecondaryColor.green
        : SecondaryColor.orange;
  } else {
    color = SecondaryColor.green;
  }

  return (
    <React.Fragment>
      <ProgressContainer
        className={classNames('aries-progress', className)}
        progress={progressValue}
        size={sizeInEm}
        role="progressbar"
        aria-valuenow={normalizedPercentage}
        aria-valuemin={0}
        aria-valuemax={100}
        tabIndex={0}
        {...defaultProps}
      >
        <ProgressContent tabIndex={-1} size={sizeInEm}>
          <svg width="8em" height="8em" viewBox="0 0 100 100">
            <circle
              cx="50"
              cy="50"
              r="45"
              fill="none"
              stroke={SecondaryColor.lighterblack}
              strokeWidth="8"
            />
            <circle
              className="progress-circle-value"
              aria-label="progress-circle-value"
              cx="50"
              cy="50"
              r="45"
              fill="none"
              stroke={color}
              strokeWidth="8"
            />
          </svg>
          <ProgressLabelWrapper aria-hidden="true">
            {content || (
              <PercentageCompletion aria-label="percentage-completion">
                {`${normalizedPercentage}%`}
              </PercentageCompletion>
            )}
            {!content && <LabelText>COMPLETE</LabelText>}
          </ProgressLabelWrapper>
        </ProgressContent>
      </ProgressContainer>
    </React.Fragment>
  );
};

interface Props
  extends Omit<
    React.ComponentPropsWithoutRef<typeof ProgressContainer>,
    'progress'
  > {
  percentage: number;
  percentageRange?: number[];
  size?: number;
  content?: React.ReactNode;
}

export default Progress;
