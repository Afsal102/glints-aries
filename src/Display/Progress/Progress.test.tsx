import * as React from 'react';
import Progress from './Progress';
import { PrimaryColor, SecondaryColor } from '../../Utils/Colors';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';

it(`<Progress> should render 51% with a stroke color of ${SecondaryColor.orange}`, () => {
  const ProgressSnapshot = renderer
    .create(<Progress percentage={51} percentageRange={[15, 60]} />)
    .toJSON();
  expect(ProgressSnapshot).toMatchSnapshot();
});

it('displays the percentage completion rate', () => {
  const percentage = 51;
  const { getByLabelText } = render(<Progress percentage={percentage} />);
  const percentageCompletion = getByLabelText('percentage-completion');
  expect(percentageCompletion).toHaveTextContent(`${percentage}%`);
});

describe('displays "0% complete" when it renders with invalid values:', () => {
  const invalidPercentages: any[] = [null, undefined, NaN];
  invalidPercentages.forEach(invalidPercentage => {
    it(`${invalidPercentage}`, () => {
      const { getByLabelText } = render(
        <Progress percentage={invalidPercentage} />
      );
      const percentageCompletion = getByLabelText('percentage-completion');
      expect(percentageCompletion).toHaveTextContent(/^0%$/);
    });
  });
});

describe('displays correct colors with a percentageRange of [15, 60]:', () => {
  const percentageRange = [15, 60];
  const [firstNumber, secondNumber] = percentageRange;
  const colorRange = {
    [firstNumber]: PrimaryColor.glintsred,
    [secondNumber]: SecondaryColor.orange,
    [secondNumber + 1]: SecondaryColor.green,
  };
  Object.keys(colorRange).forEach((range: any) => {
    const color = colorRange[range];
    it(`should display ${color} when percentage = ${range}`, () => {
      const { getByLabelText } = render(
        <Progress
          percentage={Number(range)}
          percentageRange={percentageRange}
        />
      );
      const progressCircle = getByLabelText('progress-circle-value');
      expect(progressCircle).toHaveAttribute('stroke', color);
    });
  });
});

describe('displays correct colors with a percentageRange of [40]:', () => {
  const percentageRange = [40];
  const colorRange = {
    [percentageRange[0]]: SecondaryColor.orange,
    [percentageRange[0] + 1]: SecondaryColor.green,
  };
  Object.keys(colorRange).forEach((range: any) => {
    const color = colorRange[range];
    it(`should display ${color} when percentage = ${range}`, () => {
      const { getByLabelText } = render(
        <Progress
          percentage={Number(range)}
          percentageRange={percentageRange}
        />
      );
      const progressCircle = getByLabelText('progress-circle-value');
      expect(progressCircle).toHaveAttribute('stroke', color);
    });
  });
});
