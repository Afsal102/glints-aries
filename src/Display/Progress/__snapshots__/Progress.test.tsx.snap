// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Progress> should render 51% with a stroke color of #F5A623 1`] = `
<div
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={51}
  className="ProgressStyle__ProgressContainer-sc-1fz52py-1 cNWIiW aries-progress"
  role="progressbar"
  size={1}
  tabIndex={0}
>
  <div
    className="ProgressStyle__ProgressContent-sc-1fz52py-0 kizadZ"
    size={1}
    tabIndex={-1}
  >
    <svg
      height="8em"
      viewBox="0 0 100 100"
      width="8em"
    >
      <circle
        cx="50"
        cy="50"
        fill="none"
        r="45"
        stroke="#AAAAAA"
        strokeWidth="8"
      />
      <circle
        aria-label="progress-circle-value"
        className="progress-circle-value"
        cx="50"
        cy="50"
        fill="none"
        r="45"
        stroke="#F5A623"
        strokeWidth="8"
      />
    </svg>
    <label
      aria-hidden="true"
      className="ProgressStyle__ProgressLabelWrapper-sc-1fz52py-2 jkmcgt"
    >
      <h1
        aria-label="percentage-completion"
        className="ProgressStyle__PercentageCompletion-sc-1fz52py-3 eILgds"
      >
        51%
      </h1>
      <p
        className="ProgressStyle__LabelText-sc-1fz52py-4 izBCNV"
      >
        COMPLETE
      </p>
    </label>
  </div>
</div>
`;
